File: ruby_runtime/govfuzz_driver.rb
# SPDX-License-Identifier: Apache-2.0
# govfuzz Ruby fork-server driver. A TracePoint(:line) records edge coverage into the
# shared GOVFUZZ_COV_SHM map. Speaks the GOVFUZZ_FRAMED protocol (matches
# c_runtime/govfuzz_driver.c and the Perl driver): fd 1 control pipe, fd 0 input pipe
# ({u32 LE len, bytes}). A finding raises -> stderr marker + exit 86 (no sync byte);
# an expected rejection (input validation / a library's own exception class) is
# swallowed. No third-party fuzzer.

FINDING_HALT = 86
# The declared-rejection namespace: an exception whose class is defined in the
# target's own top module (e.g. Toml::ParseError) is that library's rejection, not a
# govfuzz finding (the Ada/Java/Perl "declared exception" principle).
TARGET_MODULE = ENV['GOVFUZZ_TARGET_MODULE'].to_s
EXPECTED = (ENV['GOVFUZZ_EXPECTED_EXCEPTIONS'] || '').split(',').reject(&:empty?)

# ---- Coverage: a TracePoint(:line) folded into a 64KB AFL-style edge bitmap. ----
COV_BITS = 1 << 16
COV_MASK = COV_BITS - 1
$cov_map = ("\0".b * COV_BITS)
$cov_prev = 0
$cov_prefix = ENV['GOVFUZZ_TRACE_PREFIX'].to_s
$covered = {}
$covered_dumped = 0

def gf_fnv(str, line)
  h = 0x811c9dc5
  str.each_byte { |c| h = ((h ^ c) * 0x01000193) & 0xFFFFFFFF }
  h = ((h ^ (line & 0xFF)) * 0x01000193) & 0xFFFFFFFF
  h = ((h ^ ((line >> 8) & 0xFF)) * 0x01000193) & 0xFFFFFFFF
  h & COV_MASK
end

COV_TP = TracePoint.new(:line) do |tp|
  path = tp.path
  next if !$cov_prefix.empty? && !path.start_with?($cov_prefix)
  $covered["#{path}:#{tp.lineno}"] = true
  block = gf_fnv(path, tp.lineno)
  idx = ($cov_prev ^ block) & COV_MASK
  v = $cov_map.getbyte(idx)
  $cov_map.setbyte(idx, v + 1) if v < 0xFF
  $cov_prev = (block >> 1) & COV_MASK
end

def gf_reset_prev
  $cov_prev = 0
end

def gf_flush_cov
  path = ENV['GOVFUZZ_COV_SHM']
  return if path.nil? || path.empty?
  begin
    File.open(path, File::RDWR | File::CREAT) do |f|
      f.binmode
      f.seek(0)
      f.write($cov_map)
    end
  rescue StandardError
    # best-effort
  end
end

def gf_dump_covered
  path = ENV['GOVFUZZ_COVERED_LINES']
  return if path.nil? || path.empty?
  n = $covered.size
  return if n == $covered_dumped
  begin
    File.write("#{path}.tmp", $covered.keys.sort.join("\n"))
    File.rename("#{path}.tmp", path)
    $covered_dumped = n
  rescue StandardError
    # best-effort
  end
end

# ---- Finding classification (FP-avoidant, like the Python/Perl/JS lanes). ----
# Only high-confidence bug-class exceptions are findings; Ruby's normal input-
# rejection errors (ArgumentError/TypeError/KeyError/IndexError/NoMethodError/...) are
# suppressed. The behavioral/taint oracles (the shim) catch the security classes.
def gf_bug_cwe(err)
  case err
  when ZeroDivisionError then 'CWE-369'
  when SystemStackError then 'CWE-674' # unbounded recursion / stack exhaustion
  when NoMemoryError then 'CWE-789'    # uncontrolled memory allocation
  else
    msg = err.message.to_s
    if err.is_a?(RuntimeError) &&
       msg =~ /\b(assert(ion)?|invariant|unreachable|internal error|should (never|not) (happen|be reached|occur)|impossible|bug|corrupt)\b/i
      'CWE-617'
    end
  end
end

def gf_report_finding(cwe, msg)
  msg = msg.to_s.gsub(/\s+/, ' ').strip
  # Marker mirrors the Perl/Python/JVM drivers; the engine's parse_ruby_finding maps
  # the CWE token to a GF rule.
  $stderr.puts "== govfuzz ruby finding: #{cwe}: #{msg}"
  $stderr.flush
  exit(FINDING_HALT)
end

def gf_handle_error(err)
  cls = err.class.name.to_s
  # A blessed exception from the target's own module is a declared rejection.
  return if !TARGET_MODULE.empty? && cls.start_with?("#{TARGET_MODULE}::")
  return if EXPECTED.any? { |x| !x.empty? && err.message.to_s.include?(x) }
  cwe = gf_bug_cwe(err)
  gf_report_finding(cwe, "#{cls}: #{err.message}") if cwe
end

def gf_run_input(data)
  gf_reset_prev
  begin
    govfuzz_run_one(data)
  rescue SystemExit, SignalException
    raise
  rescue Exception => e
    gf_handle_error(e)
  end
end

def gf_read_exact(io, n)
  buf = ''.b
  while buf.bytesize < n
    chunk = io.read(n - buf.bytesize)
    return nil if chunk.nil? || chunk.empty?
    buf << chunk
  end
  buf
end

# ---- Load the target harness (defines govfuzz_run_one). ----
harness = ENV['GOVFUZZ_HARNESS'] or abort 'GOVFUZZ_HARNESS unset'
require harness

$stdin.binmode
COV_TP.enable

if ENV.key?('GOVFUZZ_FRAMED')
  # Save the control pipe (fd 1) then redirect $stdout to /dev/null so the target's
  # prints can't corrupt the sync stream (#427).
  ctl = $stdout.dup
  ctl.binmode
  ctl.sync = true
  $stdout.reopen(File.open(File::NULL, 'w'))
  ctl.write("\x01") # ready byte
  count = 0
  loop do
    hdr = gf_read_exact($stdin, 4)
    break if hdr.nil?
    n = hdr.unpack1('V')
    data = n > 0 ? gf_read_exact($stdin, n) : ''.b
    break if data.nil?
    gf_run_input(data)
    gf_flush_cov
    ctl.write("\x01") # sync byte
    count += 1
    gf_dump_covered if (count & 0x1FF).zero?
  end
else
  data =
    if ARGV[0] && File.file?(ARGV[0])
      File.binread(ARGV[0])
    else
      $stdin.read
    end
  gf_run_input(data || ''.b)
  gf_flush_cov
end

File: perl_runtime/govfuzz_driver.pl
# SPDX-License-Identifier: Apache-2.0
# govfuzz Perl fork-server driver. Run under `perl -d:GovfuzzCov` so DB::DB records
# edge coverage into the shared GOVFUZZ_COV_SHM map. Speaks the GOVFUZZ_FRAMED
# protocol (matches c_runtime/govfuzz_driver.c): fd 1 control pipe, fd 0 input pipe
# ({u32 LE len, bytes}). A finding die()s -> stderr marker + exit 86 (no sync byte);
# an expected rejection (input validation / a library's own exception object) is
# swallowed. No third-party fuzzer.
use strict;
use warnings;
require Devel::GovfuzzCov;

my $FINDING_HALT = 86;
# Top package of the target (set by the launcher): a blessed exception object whose
# class is in the target's own namespace is that library's declared rejection, not
# a govfuzz finding (the Ada/Java "declared exception" principle).
# M22: `defined(..) ? .. : ..` not the `//` defined-or operator, so this driver
# runs on Perl 5.6-5.9 (legacy gov/mil) where `//` is a syntax error (5.10+).
my $TARGET_PKG = defined($ENV{GOVFUZZ_TARGET_PACKAGE}) ? $ENV{GOVFUZZ_TARGET_PACKAGE} : '';
my @EXPECTED = grep { length } split /,/,
    (defined($ENV{GOVFUZZ_EXPECTED_EXCEPTIONS}) ? $ENV{GOVFUZZ_EXPECTED_EXCEPTIONS} : '');

my $harness = $ENV{GOVFUZZ_HARNESS} or die "GOVFUZZ_HARNESS unset\n";
require $harness;

# A die message that reads like input validation, not a bug. Mirrors the Python
# lane's ValueError/declared-exception suppression. Perl's croak-on-bad-input
# convention produces these.
# Perl uses `die` as ordinary error handling — a plain string die is almost always
# the program REJECTING bad input (the croak-on-bad-args convention), not a bug. So
# (FP-avoidant, like the Python untyped policy) we report ONLY high-confidence
# bug-class dies and suppress everything else. The behavioral/taint oracles (the
# shim) catch the security classes regardless. Returns the CWE, or '' for a generic
# die (= not a finding). A user can force-report a message via GOVFUZZ_EXPECTED... no
# — @EXPECTED here SUPPRESSES named messages (a target's known internal die text).
sub bug_cwe {
    my ($msg) = @_;
    for my $e (@EXPECTED) { return '' if length($e) && index($msg, $e) >= 0; }
    return 'CWE-674' if $msg =~ /deep recursion/i;
    return 'CWE-789' if $msg =~ /out of memory|can't allocate/i;
    return 'CWE-369' if $msg =~ /division by zero|illegal division|illegal modulus/i;
    return 'CWE-617'
        if $msg
        =~ /\b(panic|assert(ion)?|invariant|unreachable|internal error|should (never|not) (happen|be reached|occur)|impossible|corrupt)\b/i;
    return '';    # generic die = normal Perl error handling, suppressed
}

sub run_input {
    my ($data) = @_;
    Devel::GovfuzzCov::reset_prev();
    my $ok = eval {
        govfuzzgen::govfuzz_run_one($data);
        1;
    };
    return if $ok;
    my $err = $@;
    # A blessed exception object from the target's own package is a declared
    # rejection (e.g. My::Parser::Error), not a finding.
    if (ref $err) {
        my $class = ref $err;
        return if length($TARGET_PKG) && index($class, $TARGET_PKG) == 0;
        my $cwe = bug_cwe("$err");
        report_finding($cwe, "$class: $err") if $cwe;
        return;
    }
    my $cwe = bug_cwe("$err");
    report_finding($cwe, "$err") if $cwe;
}

sub report_finding {
    my ($cwe, $msg) = @_;
    $msg =~ s/\s+at .* line \d+\.?$//;    # trim the "at FILE line N" suffix
    # Marker mirrors the JVM/Python drivers; the engine's parse_perl_finding maps
    # the CWE token to a GF rule.
    print STDERR "== govfuzz perl finding: $cwe: $msg\n";
    exit($FINDING_HALT);
}

sub read_exact {
    my ($n) = @_;
    my $buf = '';
    while (length($buf) < $n) {
        my $r = sysread(STDIN, my $chunk, $n - length($buf));
        return undef if !defined($r) || $r == 0;
        $buf .= $chunk;
    }
    return $buf;
}

binmode STDIN;
if (defined $ENV{GOVFUZZ_FRAMED}) {
    # Save the control pipe (fd 1) then redirect STDOUT to /dev/null so the target's
    # prints can't corrupt the sync stream (#427).
    open(my $CTL, '>&', \*STDOUT) or die "dup control: $!\n";
    binmode $CTL;
    $CTL->autoflush(1);
    open(STDOUT, '>', '/dev/null') or die "redirect stdout: $!\n";
    syswrite($CTL, "\x01");    # ready byte
    my $count = 0;
    while (1) {
        my $hdr = read_exact(4);
        last if !defined $hdr;
        my $n = unpack('V', $hdr);
        my $data = $n > 0 ? read_exact($n) : '';
        last if !defined $data;
        run_input($data);
        Devel::GovfuzzCov::flush();
        syswrite($CTL, "\x01");    # sync byte
        # Periodically persist covered lines for negative fuzz-confirmation (no-op
        # unless the set grew / GOVFUZZ_COVERED_LINES is set; coverage plateaus).
        ++$count;
        # Exponentially spaced early checkpoints keep short campaigns from
        # losing all line evidence before the old 512-input checkpoint.
        Devel::GovfuzzCov::dump_covered_lines()
            if (($count & ($count - 1)) == 0 || ($count & 0x1FF) == 0);
    }
} else {
    my $data;
    if (@ARGV && -f $ARGV[0]) {
        local $/;
        open(my $f, '<', $ARGV[0]) or die "open $ARGV[0]: $!\n";
        binmode $f;
        $data = <$f>;
    } else {
        local $/;
        $data = <STDIN>;
    }
    run_input(defined($data) ? $data : '');
    Devel::GovfuzzCov::flush();
}
Devel::GovfuzzCov::dump_covered_lines();
